src/performance-tracking/index.ts

```diff
@@ -0,0 +1,142 @@
+  additionalParams?: AdditionalParams
+) {
+  logDurationIfAppropriate(metric, durationInMs);
+  analytics.track(metric, {
+    durationInMs,
+    ...additionalParams,
+  });
+}
+
+/**
+ * Function that starts a performance measurement.
+ * It uses performance.now() to get a start timestamp.
+ * If you need more control over timestamps or need to use another format,
+ * please use markStartTime/markFinishTime/commitMeasurement API.
+ *
+ * @param metric What you're measuring
+ * @param additionalParams Any additional context you want to add to your log
+ */
+function startMeasuring(
+  metric: PerformanceMetric,
+  additionalParams?: AdditionalParams
+) {
+  const startTime = performance.now();
+
+  currentlyTrackedMetrics.set(metric, {
+    additionalParams,
+    startTimestamp: startTime,
+  });
+}
+
+/**
+ * Function that finishes and commits a performance measurement.
+ * It uses performance.now() to get a finish timestamp.
+ * If you need more control over timestamps or need to use another format,
+ * please use markStartTime/markFinishTime/commitMeasurement API.
+ *
+ * CAUTION: Finish has to be always called after calling start for the same metric before.
+ * @param metric What you're measuring
+ * @param additionalParams Any additional context you want to add to your log
+ * @returns True if the measurement was collected and commited properly, false otherwise
+ */
+function finishMeasuring(
+  metric: PerformanceMetric,
+  additionalParams?: AdditionalParams
+): boolean {
+  const savedEntry = currentlyTrackedMetrics.get(metric);
+  if (savedEntry === undefined || savedEntry.startTimestamp === undefined) {
+    return false;
+  }
+
+  const finishTime = performance.now();
+  const durationInMs = finishTime - savedEntry.startTimestamp;
+
+  analytics.track(metric, {
+    durationInMs,
+    ...savedEntry.additionalParams,
+    ...additionalParams,
+  });
+  logDurationIfAppropriate(metric, durationInMs);
+  currentlyTrackedMetrics.delete(metric);
+  return true;
+}
+
+/**
+ * Function decorator, that tracks performance of a function using performance.now() calls
+ * and logs the result with segment.
+ * @param fn Function which performance will be measured
+ * @param metric What you're measuring, the name of the metric
+ * @param additionalParams Any additional context you want to add to your log
+ */
+export function withPerformanceTracking<Fn extends (...args: any[]) => any>(
+  fn: Fn,
+  metric: PerformanceMetric,
+  additionalParams?: AdditionalParams
+): (...args: Parameters<Fn>) => ReturnType<Fn> {
+  return function wrapper(this: any, ...args: Parameters<Fn>): ReturnType<Fn> {
+    const startTime = performance.now();
+
+    // eslint-disable-next-line babel/no-invalid-this
+    const res = fn.apply(this, args);
+
+    const durationInMs = performance.now() - startTime;
+    logDurationIfAppropriate(metric, durationInMs);
+    analytics.track(metric, { durationInMs, ...additionalParams });
+
+    return res;
+  };
+}
```

Comment:
Do you have a better idea for this? I'm not entirely sure about typing here. It works but the resulting type is not as elegant as the original one.

Reply:
What do you mean by 'original one' here?

Reply:
<img width="431" alt="image" src="https://user-images.githubusercontent.com/16062886/165324867-f414dc06-bbc3-48ce-9da2-bc45ea24e463.png">


vs


<img width="948" alt="image" src="https://user-images.githubusercontent.com/16062886/165325110-06349e20-64db-4a82-8f57-e80328456bad.png">

Reply:
It will work okay in terms of suggestions though